### PR TITLE
Switch to using GitHub Actions for CI

### PR DESCRIPTION
**Pull request by robyoung**
I have tried to keep it functionally identical to the travis workflow. This is a bit tricky for me to test.

There are some minor changes though.
- Added MSRV of 1.51.0 - I just picked a version out of thin air here.
- I could not see an easy way of preventing master branch runs for PRs that have passed so this should run on all.
- For deploying to github pages I'm using a 3rd party action. It looks reasonable and just uses the default repo scoped token.

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
+name: CI
+
+on:
+  push:
+    branches: [master, staging, trying]
+  pull_request:
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
```
Comment (adamgreig) [inline .github/workflows/ci.yaml:10]:
```suggestion
    runs-on: ubuntu-20.04
```

We've mostly specified specific versions just to ensure there are no surprises when `-latest` next changes. Same deal applies to the deploy job below, though you could also make that a final and optional step in the build job, I don't mind either way.

**Review by adamgreig — CHANGES_REQUESTED** [review]
Wow, thanks for putting this together!

On the whole this looks good but I'd like for it to be slightly more in line with our various other mdbook repos to make ongoing maintenance easier:

* https://github.com/rust-embedded/showcase/pull/29
* https://github.com/rust-embedded/book/pull/285
* https://github.com/rust-embedded/discovery/pull/311

The main things to change are:

* use cargo to install mdbook and cache it: [example](https://github.com/rust-embedded/showcase/blob/54275db8f093f072b1573bbd4e918b2b9d25a101/.github/workflows/ci.yml#L24-L42)
* use `peaceiris/actions-gh-pages@v3` for gh-pages: [example](https://github.com/rust-embedded/showcase/blob/54275db8f093f072b1573bbd4e918b2b9d25a101/.github/workflows/ci.yml#L49-L56)
* update [bors.toml](https://github.com/rust-embedded/embedonomicon/blob/master/.github/bors.toml#L4) to require the new check passes, instead of the old one ([example](https://github.com/rust-embedded/showcase/blob/54275db8f093f072b1573bbd4e918b2b9d25a101/.github/bors.toml#L4))

Hopefully those aren't too bad to do, and thanks again for the PR!

**Comment by robyoung** [comment]
That's great. Thanks for the review. I'll take a look at it later today.

**Comment by robyoung** [comment]
@adamgreig I have tried to follow the other book repos as much as possible. However, it looks like this repo does a bit more testing so it feels like it needs to be tested across different versions of rust, which the other workflows do not.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      - name: Put new bin directory into path
+        run: echo "~/cache-bin" >> $GITHUB_PATH
+
+      - name: Test
+        run: bash ci/script.sh
+        env:
+          RUST_VERSION: ${{ matrix.rust }}
+
+  deploy:
+    runs-on: ubuntu-latest
```
Comment (adamgreig) [inline .github/workflows/ci.yaml:83]:
```suggestion
    runs-on: ubuntu-20.04
```

**Review by adamgreig — CHANGES_REQUESTED** [review]
Thanks, looks good! Couple last changes then let's give it a go. Sadly since no actions have been run I don't have a button to try running them for this PR, but once the bors.toml file is updated I can have bors try it instead.

* [x] I think we can delete `ci/install.sh` and the new `ci/install-mdbook.sh` now?
* [x] Currently nightly builds failing will fail CI, but we probably want to allow nightly to fail; there's an example [here](https://github.com/rust-embedded/cortex-m/blob/master/.github/workflows/ci.yml#L11)
* [x] `.github/bors.toml` needs updating or bors will refuse to merge this PR
* [x] `.travis.yml` can be deleted too

**Review by adamgreig — DISMISSED** [review]
Great, thank you again! Let's get this merged and hopefully it works, then I'll merge #70 afterwards.

bors merge

**Comment by adamgreig** [comment]
Ugh, that's enough whack-a-mole for tonight. Please don't be put off, this is mostly the test script needing some updates for the latest version of `nm`. The usual game of fixing one CI bug and then finding the next... If you can work out why arm-none-eabi-gcc isn't being found even though it's apparently installed please go ahead, but otherwise I'll try again tomorrow.

**Comment by adamgreig** [comment]
Whew, that was an endeavour. It turns out you can't copy arm-none-eabi-gcc into a new directory, it needs to stay relative to the directory structure it was extracted from. Who knew.

bors merge

**Review by adamgreig — APPROVED** [review]
bors merge

**Comment by adamgreig** [comment]
bors cancel

**Review by adamgreig — APPROVED** [review]
bors merge

**Comment by adamgreig** [comment]
:tada: thanks @robyoung!

**Comment by robyoung** [comment]
Woop! Thank you for fixing it. I was totally stumped.

**Pull request merged**